crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,93 @@
+//! Wrapper for [`discv5::Discv5`] that supports downgrade to [`Discv4`].
+
+use std::{
+    fmt,
+    pin::Pin,
+    sync::Arc,
+    task::{ready, Context, Poll},
+};
+
+use derive_more::From;
+use futures::{Stream, StreamExt};
+use parking_lot::RwLock;
+use reth_discv4::{DiscoveryUpdate, Discv4, HandleDiscovery};
+use tokio::sync::{mpsc, watch};
+use tokio_stream::{wrappers::ReceiverStream, StreamExt as TokioStreamExt};
+
+/// Reth Discv5 type, wraps [`discv5::Discv5`] supporting downgrade to [`Discv4`].
+pub struct Discv5 {
+    _discv5: Arc<RwLock<discv5::Discv5>>,
+    discv4: Discv4,
+    discv5_kbuckets_change_tx: watch::Sender<()>,
+}
+
+impl Discv5 {
+    /// Returns a new [`Discv5`] handle.
+    pub fn new(
+        discv5: Arc<RwLock<discv5::Discv5>>,
+        discv4: Discv4,
+        discv5_kbuckets_change_tx: watch::Sender<()>,
+    ) -> Self {
+        Self { _discv5: discv5, discv4, discv5_kbuckets_change_tx }
+    }
+
+    /// Notifies [`Discv4`] that [discv5::Discv5]'s kbucktes have been updated. This brings
+    /// [`Discv4`] to update its mirror of the [discv5::Discv5] kbucktes upon next
+    /// [`reth_discv4::Neighbours`] message.
+    pub fn notify_discv4_of_kbuckets_update(&self) -> Result<(), watch::error::SendError<()>> {
+        self.discv5_kbuckets_change_tx.send(())
+    }
+}
+
+impl HandleDiscovery for Discv5 {
+    // todo: delegate methods to both discv5 and discv4
+}
+
+impl fmt::Debug for Discv5 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut debug_struct = f.debug_struct("Discv5");
+
+        debug_struct.field("discv5", &"{ .. }");
+        debug_struct.field("discv4", &self.discv4);
+        debug_struct.field("discv5_kbuckets_change_tx", &self.discv5_kbuckets_change_tx);
+
+        debug_struct.finish()
+    }
+}
+
+/// Wrapper around update type used in [`discv5::Discv5`] and [`Discv4`].
+#[derive(Debug, From)]
+pub enum DiscoveryUpdateV5 {
+    /// A [`discv5::Discv5`] update.
+    V5(discv5::Discv5Event),
+    /// A [`Discv4`] update.
+    V4(DiscoveryUpdate),
+}
+
+/// Stream wrapper for streams producing types that can convert to [`DiscoveryUpdateV5`].
+#[derive(Debug)]
+pub struct UpdateStream<S>(S);
+
+impl<S, I> Stream for UpdateStream<S>
+where
+    S: Stream<Item = I> + Unpin,
+    DiscoveryUpdateV5: From<I>,
+{
+    type Item = DiscoveryUpdateV5;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Poll::Ready(ready!(self.0.poll_next_unpin(cx)).map(DiscoveryUpdateV5::from))
+    }
+}
+
+/// Returns a merged stream for  [`discv5::Discv5Event`] stream, that supports downgrading to
+/// discv4.
+pub fn merge_discovery_streams(
+    discv5_event_stream: mpsc::Receiver<discv5::Discv5Event>,
+    discv4_update_stream: ReceiverStream<DiscoveryUpdate>,
+) -> impl Stream<Item = DiscoveryUpdateV5> + Unpin {
```

Comment:
Tokio doesn't publicly expose the type `Merge` unfortunately. if we want to avoid storing the the merged discovery updates stream as a trait object in `NetworkState`, we will have to implement our own type like `Merge`.
```rust
#[pin_project]
pub struct MergedStream {
    #[pin]
    discv5_update_stream: UpdateStream<ReceiverStream<discv5::Discv5Event>>,
    #[pin]
    discv4_update_stream: UpdateStream<ReceiverStream<DiscoveryUpdate>>,
    should_poll_discv5_first: bool,
}

impl Stream for MergedStream {
    type Item = DiscoveryUpdateV5;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let mut this = self.project();
        if *this.should_poll_discv5_first {
            *this.should_poll_discv5_first = false;
            let discv5_res = this.discv5_update_stream.poll_next_unpin(cx);
            if discv5_res.is_pending() {
                return this.discv4_update_stream.poll_next_unpin(cx);
            }
            return discv5_res
        }

        *this.should_poll_discv5_first = true;
        let discv4_res = this.discv4_update_stream.poll_next_unpin(cx);
        if discv4_res.is_pending() {
            return this.discv5_update_stream.poll_next_unpin(cx);
        }
        discv4_res
    }
}
```
your call @mattsse :)

Reply:
isn't this just `stream::Select` ?

Reply:
more or less, `select` uses random to select one if both are ready. `merge` alternates fairly.